File: backend/app/routes/execute.py
"""Script execution route: runs a generated automation script server-side.

The script executes in a sandboxed subprocess against the twin — the exact
code a user would deploy on a real instrument. Events (logs, images, errors)
stream back over SSE. One run at a time: while a run is active, mutating
microscope/simulation endpoints return 409.
"""

import json
from typing import Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from ..services import script_runner
from ..services import twin_session as ts

router = APIRouter(prefix="/execute", tags=["execute"])


class RunScriptRequest(BaseModel):
    code: str
    timeout_s: int = script_runner.DEFAULT_TIMEOUT_S
    label: Optional[str] = None


@router.get("/status")
def get_run_status():
    return ts.run_status()


@router.post("/run")
def run_script(request: RunScriptRequest):
    """Execute the script, streaming events as SSE. 409 if a run is active."""
    if not request.code.strip():
        raise HTTPException(status_code=400, detail="Script is empty.")
    timeout_s = max(5, min(request.timeout_s, 1800))

    if not ts.try_begin_run(label=request.label or "script"):
        raise HTTPException(
            status_code=409,
            detail="A script run is already in progress.",
        )

    def event_stream():
        try:
            for event in script_runner.run_script(request.code, timeout_s=timeout_s):
                yield f"data: {json.dumps(event)}\n\n"
        finally:
            ts.end_run()

    return StreamingResponse(event_stream(), media_type="text/event-stream")

File: src/api/execute.ts
/**
 * API client for server-side script execution (/api/execute).
 *
 * The generated Python runs in a sandboxed subprocess on the backend — the
 * exact script a user would deploy on a real instrument. Events stream back
 * over SSE: logs, acquired frames (via the image-marker protocol), errors,
 * and a final done event.
 */

import { API_BASE_URL, ApiError } from './client';

export interface RunLogEvent { type: 'log'; message: string }
export interface RunErrorEvent { type: 'error'; message: string }
export interface RunImageEvent {
  type: 'image';
  image: { image_base64: string; width: number; height: number; dtype: string };
  meta: Record<string, unknown>;
}
export interface RunDoneEvent {
  type: 'done';
  exit_code: number;
  elapsed_s: number;
  images: number;
}
export type RunEvent = RunLogEvent | RunErrorEvent | RunImageEvent | RunDoneEvent;

export interface RunStatus {
  active: boolean;
  started_at: number | null;
  label: string | null;
}

export async function getRunStatus(): Promise<RunStatus> {
  const response = await fetch(`${API_BASE_URL}/execute/status`);
  if (!response.ok) throw new ApiError(response.status, 'Failed to get run status');
  return response.json();
}

/**
 * Run a script, invoking `onEvent` for each streamed event.
 * Resolves when the stream ends; rejects with ApiError on start failure
 * (409 = a run is already active). Abort via the optional AbortSignal.
 */
export async function runScript(
  code: string,
  onEvent: (event: RunEvent) => void,
  options: { timeoutS?: number; label?: string; signal?: AbortSignal } = {},
): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/execute/run`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      code,
      timeout_s: options.timeoutS ?? 300,
      label: options.label ?? null,
    }),
    signal: options.signal,
  });

  if (!response.ok || !response.body) {
    let detail = `Run failed to start (${response.status})`;
    try {
      const body = await response.json();
      if (typeof body.detail === 'string') detail = body.detail;
    } catch {
      // keep generic detail
    }
    throw new ApiError(response.status, detail);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    // SSE frames are separated by a blank line.
    let sep: number;
    while ((sep = buffer.indexOf('\n\n')) >= 0) {
      const frame = buffer.slice(0, sep);
      buffer = buffer.slice(sep + 2);
      for (const line of frame.split('\n')) {
        if (!line.startsWith('data: ')) continue;
        try {
          onEvent(JSON.parse(line.slice(6)) as RunEvent);
        } catch {
          // Malformed frame — surface rather than swallow.
          onEvent({ type: 'error', message: 'Received malformed event from server' });
        }
      }
    }
  }
}
